### PR TITLE
storage: retry cleaning up rocksdb db's on startup

### LINKED ISSUE #19999: IO error: lock hold by current process, acquire time 1686932018 acquiring thread 139789160330816: /mzdata/source_data/u46/3/LOCK: No locks available
### What version of Materialize are you using?

8550fe6

### What is the issue?

```
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: Error { message: "IO error: lock hold by current process, acquire time 1686932018 acquiring thread 139789160330816: /mzdata/source_data/u46/3/LOCK: No locks available" }', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-057cba75ebdb1f2c8-1/materialize/nightlies/src/rocksdb/src/lib.rs:291:82 
```
https://buildkite.com/materialize/nightlies/builds/2614#_ (also retriggered to see if flaky or actually coming from that change)

I haven't seen these failures before today. They happened in @sploiselle's change fixing another panic: https://github.com/MaterializeInc/materialize/pull/19996
But seem to be unrelated to the change. I'm guessing they are flakes? I didn't see any RocksDB-related changes recently.

ci-regexp: IO error: lock hold by current process, acquire time

### PR DESCRIPTION
**Pull request by guswynn**
Possibly resolves #19999 and #25417

The past way of doing this ALLOWED us to skip errors on destruction of the db, and then panic on creation of the db; this is not only less flaky (we hope, at least in my local testing it isnt as flaky), but also safer, with us never trying to build a db before successfully deleting it.

### Motivation
  * This PR fixes a recognized bug.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/rocksdb/src/lib.rs [hunk 1]
```diff
@@ -244,48 +249,55 @@ where
         let dynamic_config = tuning_config.dynamic.clone();
         if options.cleanup_on_new && instance_path.exists() {
```
Comment (benesch) [inline src/rocksdb/src/lib.rs:250]: Why is `cleanup_on_new` an option? It seems like it's true always?

File: src/rocksdb/src/lib.rs [hunk 2]
```diff
@@ -244,48 +249,55 @@ where
...
-            let legacy_instance_path_owned = legacy_instance_path.to_owned();
-            mz_ore::task::spawn_blocking(
-                || {
-                    format!(
-                        "RocksDB instance at {}: cleanup on creation",
-                        instance_path.display()
+
+            // We require that cleanup of the DB succeeds. Otherwise, we could open a DB with old,
+            // incorrect data. Because of races with dataflow shutdown, we retry a few times here.
+            // 1s with a 2x backoff is ~30s after 5 tries.
```
Comment (benesch) [inline src/rocksdb/src/lib.rs:255]: Can you add a TODO here that indicates this entire thing could and should go away when the controller can await the dataflow's destruction before restarting it?

**Comment by sploiselle** [comment]
@guswynn Can we get in a few runs of the failing `KafkaParallelInsert` zippy test for anecdotal evidence this moves the needle?

**Review by benesch — APPROVED** [review]
This looks "good" to me. The raciness here makes me uncomfortable, but I can't come up with an edge case that this approach fails to handle.

**Comment by maddyblue** [comment]
Adapter part lgtm.

**Comment by guswynn** [comment]
@sploiselle will do!

File: src/rocksdb/src/lib.rs (hunk 1 above)
Comment (guswynn) [reply src/rocksdb/src/lib.rs:250]: I had this very early on and just never removed it! I can cleanup in a followup

File: src/rocksdb/src/lib.rs (hunk 2 above)
Comment (guswynn) [reply src/rocksdb/src/lib.rs:255]: yes definitely!

**Comment by benesch** [comment]
@guswynn you probably want to remove these as well:

  * https://github.com/MaterializeInc/materialize/blob/e5767ba1fc6c60c30de5791628d1523fff97e9f7/ci/release-qualification/pipeline.template.yml#L130
  * https://github.com/MaterializeInc/materialize/blob/e5767ba1fc6c60c30de5791628d1523fff97e9f7/misc/python/materialize/zippy/source_actions.py#L131-L133

I'd suggest removing the check that prevents altering connections that have upsert sources attached in this PR, so that we can really put this new code path through its paces.

**Comment by guswynn** [comment]
Yeah @benesch I couldn't come up with anything better to handle the race!

I re-enabled all the tests and will run the nightly tests a few times, and the rq tests

**Comment by guswynn** [comment]
RQ: https://buildkite.com/materialize/release-qualification/builds/488
nightly: https://buildkite.com/materialize/nightly/builds/7567#_

**Comment by guswynn** [comment]
This passed the `KafkaParallelInsert` zippy twice, and it re-enables `Longer Zippy Kafka Parallel Insert`, which is running tonight here: https://buildkite.com/materialize/release-qualification/builds/488

I am going to go ahead and handle the rebase+merge conflicts and merge!

**Comment by benesch** [comment]
> I am going to go ahead and handle the rebase+merge conflicts and merge!

Nice! Thank you, Gus!

**Comment by guswynn** [comment]
RQ's failed with some docker error: https://buildkite.com/materialize/release-qualification/builds/488#018f1bdb-a788-4456-ba22-0eabebec4169

**Pull request merged**